lib/vrl/stdlib/src/get_host_ip.rs

```diff
@@ -0,0 +1,100 @@
+                required: false,
+            },
+            Parameter {
+                keyword: "family",
+                kind: kind::BYTES,
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let families = vec![value!("IPv4"), value!("IPv6")];
+        let interface = arguments.optional("interface");
+        let family = arguments
+            .optional_enum("variant", &families)?
+            .map(|v| v.try_bytes().expect("variant not bytes"));
+        Ok(Box::new(GetHostIpFn { interface, family }))
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "valid",
+                source: r#"get_host_ip!() != null"#,
+                result: Ok("true"),
+            },
+            Example {
+                title: "IPv4",
+                source: r#"get_host_ip!(family: "IPv4") != null"#,
+                result: Ok("true"),
+            },
+            Example {
+                title: "interface",
+                source: r#"get_host_ip!(interface: "eth0") != null"#,
+                result: Ok("true"),
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct GetHostIpFn {
+    interface: Option<Box<dyn Expression>>,
+    family: Option<Bytes>,
+}
+
+impl Expression for GetHostIpFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let interface = self
+            .interface
+            .as_ref()
+            .map(|d| d.resolve(ctx))
+            .transpose()?
+            .map(|d| d.try_bytes())
+            .transpose()?;
+
+        // This attempts to just find the first non-loopback
+        // interface that is up and grab its first IP
+        let default_ip = pnet_datalink::interfaces()
+            .iter()
+            // If user specified a interface, find that one, otherwise find the first non-loopback
+            // interface that is up
+            .find(|e| match &interface {
+                Some(interface) => e.name.as_bytes() == interface,
+                None => e.is_up() && !e.is_loopback() && !e.ips.is_empty(),
+            })
+            // If the user specifid an address family, find the first address matching that family,
+            // otherwise find the first ip
+            .and_then(|interface| match self.family.as_ref().map(|b| b.as_ref()) {
+                Some(b"IPv4") => interface.ips.iter().find(|ip| ip.is_ipv4()),
+                Some(b"IPv6") => interface.ips.iter().find(|ip| ip.is_ipv6()),
+                None => interface.ips.get(0),
+                _ => unreachable!("enum invariant"),
+            })
+            .map(|ip| ip.ip().to_string());
+
+        default_ip
+            .map(Into::into)
+            .ok_or_else(|| "unable to find IP address".into())
+    }
+
+    fn type_def(&self, _: &state::Compiler) -> TypeDef {
+        TypeDef::new().fallible().bytes().add_null()
```

Comment:
```suggestion
        TypeDef::new().fallible().bytes()
```

I don't believe this can return `null` any more, can it?

Reply:
Probably good to add a couple of unit tests which also validate the returned type def.

(not sure if the `function_tests` macro can be used here, if not, regular unit tests would still be good to add)

Reply:
👍 I added a test that fetches the loopback address using `test_function`, which I think should be pretty safe (I have it only run on `cfg(unix)`). I also added another plain unit tests that verifies it can fetch _an_ address.